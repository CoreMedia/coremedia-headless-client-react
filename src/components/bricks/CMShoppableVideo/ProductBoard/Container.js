// @flow
// import PropTypes from 'prop-types';
import styled from 'styled-components';

import { media } from '../../../../styles/themes/utils';

const Container = styled.div`
  display: flex;
  flex-wrap: nowrap;
  align-items: center;
  height: 100%;
  transition: transform 0.45s ease-out;
  ${props => media(props.theme.breakpoints.tablet)`
    flex-wrap: wrap;
    flex-direction: row;
    transform: unset;
    justify-content: center;
    align-content: ${props.productboardOverflow ? 'flex-start' : 'center'};
  `};
`;

Container.displayName = 'Container';

/*Container.propTypes = {
  theme: PropTypes.shape({
    breakpoints: PropTypes.shape({
      tablet: PropTypes.number.isRequired,
    }).isRequired,
  }),
};*/

Container.defaultProps = {
  theme: {
    breakpoints: {
      tablet: 768,
    },
  },
};

export default Container;
