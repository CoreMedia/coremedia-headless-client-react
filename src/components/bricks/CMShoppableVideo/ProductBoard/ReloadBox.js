// @flow
// import PropTypes from 'prop-types';
import styled from 'styled-components';

import { media } from '../../../../styles/themes/utils';

const ReloadBox = styled.div`
  display: none;
  ${props => media(props.theme.breakpoints.tablet)`
    display: block;
    position: absolute;
    bottom: 3px;
    left: 0;
    z-index: 9999;
  `};
`;

ReloadBox.displayName = 'ReloadBox';

/*ReloadBox.propTypes = {
  theme: PropTypes.shape({
    breakpoints: PropTypes.shape({
      tablet: PropTypes.number.isRequired,
    }).isRequired,
  }),
};*/

ReloadBox.defaultProps = {
  theme: {
    breakpoints: {
      tablet: 768,
    },
  },
};

export default ReloadBox;
