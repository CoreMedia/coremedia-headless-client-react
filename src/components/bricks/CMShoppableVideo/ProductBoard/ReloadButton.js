// @flow
import React from 'react';
import PropTypes from 'prop-types';

import { IconButton as BasicIconButton } from '../../../basic/Button';
import RepeatIcon from '../../../basic/Icon/RepeatIcon';
import { media } from '../../../../styles/themes/utils';

const IconButton = BasicIconButton.extend`
  width: 100%;
  height: auto;
  ${props => media(props.theme.breakpoints.tablet)`
    width: 40px;
    height: 40px;
  `};
`;

IconButton.displayName = 'IconButton';

IconButton.propTypes = {
  theme: PropTypes.shape({
    breakpoints: PropTypes.shape({
      tablet: PropTypes.number.isRequired,
    }).isRequired,
  }),
};

IconButton.defaultProps = {
  theme: {
    breakpoints: {
      tablet: 768,
    },
  },
};

type Props = {
  id?: string,
  handleClick: (ev: SyntheticEvent<HTMLButtonElement>) => void,
};

const ReloadButton = ({ id, handleClick }: Props) => (
  <IconButton onClick={handleClick} color="#808080" backgroundColor="rgba(255, 255, 255, 0.5)">
    <RepeatIcon id={id} />
  </IconButton>
);

ReloadButton.propTypes = {
  id: PropTypes.string,
  handleClick: PropTypes.func.isRequired,
};

export default ReloadButton;
