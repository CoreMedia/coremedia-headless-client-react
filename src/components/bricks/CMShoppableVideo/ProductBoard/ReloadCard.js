// @flow
// import PropTypes from 'prop-types';
import styled from 'styled-components';

import { media } from '../../../../styles/themes/utils';

const ProductBox = styled.div`
  display: inline-flex;
  flex: 0 0 auto;
  position: relative;
  width: 144px;
  height: 173px;
  margin: 8px;
  padding: 0 36px;
  order: ${props => props.index};
  ${props => media(props.theme.breakpoints.tablet)`
    display: none;
  `};
`;

ProductBox.displayName = 'ProductBox';

/*ProductBox.propTypes = {
  theme: PropTypes.shape({
    breakpoints: PropTypes.shape({
      tablet: PropTypes.number.isRequired,
    }).isRequired,
  }),
};*/

ProductBox.defaultProps = {
  theme: {
    breakpoints: {
      tablet: 768,
      desktop: 992,
      large: 1280,
    },
  },
};

export default ProductBox;
