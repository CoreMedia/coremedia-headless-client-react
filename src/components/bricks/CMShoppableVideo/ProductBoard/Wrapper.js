// @flow
// import PropTypes from 'prop-types';
import styled, { keyframes } from 'styled-components';
import theme from 'styled-theming';

import { media } from '../../../../styles/themes/utils';

const borderColor = theme('name', {
  default: 'rgba(0, 108, 174, 1)',
  hkm: '#fabfc0',
});

const fadeIn = keyframes`
  from {
    opacity 0;
  }

  to {
    opacity 1;
  }
`;

const ProductBoardWrapper = styled.div`
  position: absolute;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  background-color: #fff;
  overflow-x: scroll;
  overflow-y: hidden;
  direction: rtl;
  border-width: 3px 0;
  border-style: solid none;
  border-color: ${borderColor};
  padding: 7px;
  animation: ${fadeIn} 0.6s linear;
  ${props => media(props.theme.breakpoints.tablet)`
    display: initial;
    overflow-x: hidden;
    overflow-y: scroll;
    direction: ltr;
  `};
`;

ProductBoardWrapper.displayName = 'ProductBoardWrapper';

/*ProductBoardWrapper.propTypes = {
  theme: PropTypes.shape({
    breakpoints: PropTypes.shape({
      tablet: PropTypes.number.isRequired,
    }).isRequired,
    name: PropTypes.string.isRequired,
  }),
};*/

ProductBoardWrapper.defaultProps = {
  theme: {
    breakpoints: {
      tablet: 768,
    },
    name: 'default',
  },
};

export default ProductBoardWrapper;
