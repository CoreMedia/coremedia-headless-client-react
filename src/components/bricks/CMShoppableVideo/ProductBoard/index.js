// @flow
import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { withTheme } from 'styled-components';
import { Transition, TransitionGroup } from 'react-transition-group';
import Media from 'react-media';

import ProductCard from '../ProductCard';
import Wrapper from './Wrapper';
import Container from './Container';
import ReloadCard from './ReloadCard';
import ReloadBox from './ReloadBox';
import ReloadButton from './ReloadButton';

type Props = {
  thumbnails: Array<Object>,
  productboardOverflow: boolean,
  handleReplay: () => void,
  handleRef: (element: ?HTMLElement) => HTMLElement | null | void,
  theme: Object,
};

const ProductBoard = ({
  thumbnails,
  productboardOverflow,
  handleReplay,
  handleRef,
  theme,
}: Props) => {
  return (
    <Fragment>
      <ReloadBox>
        <ReloadButton handleClick={handleReplay} />
      </ReloadBox>
      <Wrapper>
        <Media query={`(min-width: ${theme.breakpoints.tablet}px)`}>
          {matches =>
            matches ? (
              <TransitionGroup
                component={Container}
                innerRef={handleRef}
                productboardOverflow={productboardOverflow}
              >
                {thumbnails.map((item, index) => (
                  <Transition key={item.id} appear={true} timeout={300 + index * 200}>
                    {status => (
                      <ProductCard
                        index={index}
                        entered={status === 'entered'}
                        handleClick={item.handleClick}
                        pictureLink={item.pictureLink}
                        pictureTitle={item.pictureTitle}
                        pictureAlt={item.pictureAlt}
                        title={item.teaserTitle}
                        price={item.price}
                      />
                    )}
                  </Transition>
                ))}
              </TransitionGroup>
            ) : (
              <TransitionGroup component={Container} innerRef={handleRef}>
                {thumbnails.reverse().map((item, index) => (
                  <Transition key={item.id} appear={true} timeout={300 + index * 200}>
                    {status => (
                      <ProductCard
                        index={index + 1}
                        entered={status === 'entered'}
                        handleClick={item.handleClick}
                        pictureLink={item.pictureLink}
                        pictureTitle={item.pictureTitle}
                        pictureAlt={item.pictureAlt}
                        title={item.teaserTitle}
                        price={item.price}
                      />
                    )}
                  </Transition>
                ))}
                <ReloadCard index={0}>
                  <ReloadButton handleClick={handleReplay} />
                </ReloadCard>
              </TransitionGroup>
            )
          }
        </Media>
      </Wrapper>
    </Fragment>
  );
};

const Wrapped = withTheme(ProductBoard);

Wrapped.propTypes = {
  thumbnails: PropTypes.arrayOf(PropTypes.object).isRequired,
  productboardOverflow: PropTypes.bool.isRequired,
  handleReplay: PropTypes.func.isRequired,
  handleRef: PropTypes.func.isRequired,
};

export default Wrapped;
