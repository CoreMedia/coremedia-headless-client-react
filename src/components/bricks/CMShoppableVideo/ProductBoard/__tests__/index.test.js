// @flow
import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

import ProductBoard from '../';

const createMockMediaMatcher = matches => () => ({
  matches,
  addListener: () => {},
  removeListener: () => {},
});

describe('ProductBoard Component', () => {
  let originalMatchMedia;
  beforeEach(() => {
    originalMatchMedia = window.matchMedia;
  });

  afterEach(() => {
    window.matchMedia = originalMatchMedia;
  });

  it('should render correctly for mobile devices', () => {
    window.matchMedia = createMockMediaMatcher(false);

    const thumbnails = [
      {
        id: 'ID1',
        teaserTitle: 'Title 1',
        teaserText: 'Text 1',
        pictureLink: 'Link 1',
        pictureTitle: 'Picture Title 1',
        pictureAlt: 'Picture Alt 1',
        price: 100,
        handleClick: jest.fn().mockName('handleClick'),
      },
      {
        id: 'ID2',
        teaserTitle: 'Title 2',
        teaserText: 'Text 2',
        pictureLink: 'Link 2',
        pictureTitle: 'Picture Title 2',
        pictureAlt: 'Picture Alt 2',
        price: 100,
        handleClick: jest.fn().mockName('handleClick'),
      },
      {
        id: 'ID3',
        teaserTitle: 'Title 3',
        teaserText: 'Text 3',
        pictureLink: 'Link 3',
        pictureTitle: 'Picture Title 3',
        pictureAlt: 'Picture Alt 3',
        price: 100,
        handleClick: jest.fn().mockName('handleClick'),
      },
    ];
    const handleReplay = jest.fn().mockName('handleReplay');
    const handleRef = jest.fn().mockName('handleRef');

    const wrapper = shallow(
      <ProductBoard
        thumbnails={thumbnails}
        handleReplay={handleReplay}
        handleRef={handleRef}
        productboardOverflow={false}
        theme={{
          breakpoints: {
            small: 480,
            tablet: 768,
            desktop: 992,
            large: 1280,
          },
        }}
      />
    );
    const boardWrapper = wrapper.find('ProductBoard').shallow();
    expect(boardWrapper).toMatchSnapshot();
    const mediaWrapper = boardWrapper.find('Media').shallow();
    expect(mediaWrapper).toMatchSnapshot();
    const transitionWrapper = mediaWrapper.find('Transition');
    expect(transitionWrapper).toHaveLength(3);
    transitionWrapper.forEach((node, index) => {
      expect(node.shallow()).toMatchSnapshot();
    });
  });

  it('should render correctly for tablet and larger devices', () => {
    window.matchMedia = createMockMediaMatcher(true);

    const thumbnails = [
      {
        id: 'ID1',
        teaserTitle: 'Title 1',
        teaserText: 'Text 1',
        pictureLink: 'Link 1',
        pictureTitle: 'Picture Title 1',
        pictureAlt: 'Picture Alt 1',
        price: 100,
        handleClick: jest.fn().mockName('handleClick'),
      },
      {
        id: 'ID2',
        teaserTitle: 'Title 2',
        teaserText: 'Text 2',
        pictureLink: 'Link 2',
        pictureTitle: 'Picture Title 2',
        pictureAlt: 'Picture Alt 2',
        price: 100,
        handleClick: jest.fn().mockName('handleClick'),
      },
      {
        id: 'ID3',
        teaserTitle: 'Title 3',
        teaserText: 'Text 3',
        pictureLink: 'Link 3',
        pictureTitle: 'Picture Title 3',
        pictureAlt: 'Picture Alt 3',
        price: 100,
        handleClick: jest.fn().mockName('handleClick'),
      },
    ];
    const handleReplay = jest.fn().mockName('handleReplay');
    const handleRef = jest.fn().mockName('handleRef');

    const wrapper = shallow(
      <ProductBoard
        thumbnails={thumbnails}
        handleReplay={handleReplay}
        handleRef={handleRef}
        productboardOverflow={false}
        theme={{
          breakpoints: {
            small: 480,
            tablet: 768,
            desktop: 992,
            large: 1280,
          },
        }}
      />
    );
    const boardWrapper = wrapper.find('ProductBoard').shallow();
    expect(boardWrapper).toMatchSnapshot();
    const mediaWrapper = boardWrapper.find('Media').shallow();
    expect(mediaWrapper).toMatchSnapshot();
    const transitionWrapper = mediaWrapper.find('Transition');
    expect(transitionWrapper).toHaveLength(3);
    transitionWrapper.forEach((node, index) => {
      expect(node.shallow()).toMatchSnapshot();
    });
  });
});
