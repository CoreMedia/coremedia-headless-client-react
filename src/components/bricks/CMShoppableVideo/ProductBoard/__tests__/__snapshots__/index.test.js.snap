// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductBoard Component should render correctly for mobile devices 1`] = `
<React.Fragment>
  <ReloadBox
    theme={
      Object {
        "breakpoints": Object {
          "tablet": 768,
        },
      }
    }
  >
    <ReloadButton
      handleClick={[MockFunction handleReplay]}
    />
  </ReloadBox>
  <ProductBoardWrapper
    theme={
      Object {
        "breakpoints": Object {
          "tablet": 768,
        },
        "name": "default",
      }
    }
  >
    <Media
      defaultMatches={true}
      query="(min-width: 768px)"
    />
  </ProductBoardWrapper>
</React.Fragment>
`;

exports[`ProductBoard Component should render correctly for mobile devices 2`] = `
<TransitionGroup
  childFactory={[Function]}
  component={[Function]}
  innerRef={[MockFunction handleRef]}
>
  <Transition
    appear={true}
    enter={true}
    exit={true}
    in={false}
    key="ID3"
    mountOnEnter={false}
    onEnter={[Function]}
    onEntered={[Function]}
    onEntering={[Function]}
    onExit={[Function]}
    onExited={[Function]}
    onExiting={[Function]}
    timeout={300}
    unmountOnExit={false}
  />
  <Transition
    appear={true}
    enter={true}
    exit={true}
    in={false}
    key="ID2"
    mountOnEnter={false}
    onEnter={[Function]}
    onEntered={[Function]}
    onEntering={[Function]}
    onExit={[Function]}
    onExited={[Function]}
    onExiting={[Function]}
    timeout={500}
    unmountOnExit={false}
  />
  <Transition
    appear={true}
    enter={true}
    exit={true}
    in={false}
    key="ID1"
    mountOnEnter={false}
    onEnter={[Function]}
    onEntered={[Function]}
    onEntering={[Function]}
    onExit={[Function]}
    onExited={[Function]}
    onExiting={[Function]}
    timeout={700}
    unmountOnExit={false}
  />
  <ProductBox
    index={0}
    theme={
      Object {
        "breakpoints": Object {
          "desktop": 992,
          "large": 1280,
          "tablet": 768,
        },
      }
    }
  >
    <ReloadButton
      handleClick={[MockFunction handleReplay]}
    />
  </ProductBox>
</TransitionGroup>
`;

exports[`ProductBoard Component should render correctly for mobile devices 3`] = `
<ProductCard
  entered={false}
  handleClick={[MockFunction handleClick]}
  index={1}
  pictureAlt="Picture Alt 3"
  pictureLink="Link 3"
  pictureTitle="Picture Title 3"
  price={100}
  title="Title 3"
/>
`;

exports[`ProductBoard Component should render correctly for mobile devices 4`] = `
<ProductCard
  entered={false}
  handleClick={[MockFunction handleClick]}
  index={2}
  pictureAlt="Picture Alt 2"
  pictureLink="Link 2"
  pictureTitle="Picture Title 2"
  price={100}
  title="Title 2"
/>
`;

exports[`ProductBoard Component should render correctly for mobile devices 5`] = `
<ProductCard
  entered={false}
  handleClick={[MockFunction handleClick]}
  index={3}
  pictureAlt="Picture Alt 1"
  pictureLink="Link 1"
  pictureTitle="Picture Title 1"
  price={100}
  title="Title 1"
/>
`;

exports[`ProductBoard Component should render correctly for tablet and larger devices 1`] = `
<React.Fragment>
  <ReloadBox
    theme={
      Object {
        "breakpoints": Object {
          "tablet": 768,
        },
      }
    }
  >
    <ReloadButton
      handleClick={[MockFunction handleReplay]}
    />
  </ReloadBox>
  <ProductBoardWrapper
    theme={
      Object {
        "breakpoints": Object {
          "tablet": 768,
        },
        "name": "default",
      }
    }
  >
    <Media
      defaultMatches={true}
      query="(min-width: 768px)"
    />
  </ProductBoardWrapper>
</React.Fragment>
`;

exports[`ProductBoard Component should render correctly for tablet and larger devices 2`] = `
<TransitionGroup
  childFactory={[Function]}
  component={[Function]}
  innerRef={[MockFunction handleRef]}
  productboardOverflow={false}
>
  <Transition
    appear={true}
    enter={true}
    exit={true}
    in={false}
    key="ID1"
    mountOnEnter={false}
    onEnter={[Function]}
    onEntered={[Function]}
    onEntering={[Function]}
    onExit={[Function]}
    onExited={[Function]}
    onExiting={[Function]}
    timeout={300}
    unmountOnExit={false}
  />
  <Transition
    appear={true}
    enter={true}
    exit={true}
    in={false}
    key="ID2"
    mountOnEnter={false}
    onEnter={[Function]}
    onEntered={[Function]}
    onEntering={[Function]}
    onExit={[Function]}
    onExited={[Function]}
    onExiting={[Function]}
    timeout={500}
    unmountOnExit={false}
  />
  <Transition
    appear={true}
    enter={true}
    exit={true}
    in={false}
    key="ID3"
    mountOnEnter={false}
    onEnter={[Function]}
    onEntered={[Function]}
    onEntering={[Function]}
    onExit={[Function]}
    onExited={[Function]}
    onExiting={[Function]}
    timeout={700}
    unmountOnExit={false}
  />
</TransitionGroup>
`;

exports[`ProductBoard Component should render correctly for tablet and larger devices 3`] = `
<ProductCard
  entered={false}
  handleClick={[MockFunction handleClick]}
  index={0}
  pictureAlt="Picture Alt 1"
  pictureLink="Link 1"
  pictureTitle="Picture Title 1"
  price={100}
  title="Title 1"
/>
`;

exports[`ProductBoard Component should render correctly for tablet and larger devices 4`] = `
<ProductCard
  entered={false}
  handleClick={[MockFunction handleClick]}
  index={1}
  pictureAlt="Picture Alt 2"
  pictureLink="Link 2"
  pictureTitle="Picture Title 2"
  price={100}
  title="Title 2"
/>
`;

exports[`ProductBoard Component should render correctly for tablet and larger devices 5`] = `
<ProductCard
  entered={false}
  handleClick={[MockFunction handleClick]}
  index={2}
  pictureAlt="Picture Alt 3"
  pictureLink="Link 3"
  pictureTitle="Picture Title 3"
  price={100}
  title="Title 3"
/>
`;
